src/aktualizr_primary/secondary.cc

```diff
@@ -141,20 +147,53 @@ class SecondaryWaiter {
   std::unordered_set<std::string> secondaries_to_wait_for_;
 };
 
-static Secondaries createIPSecondaries(const IPSecondariesConfig& config) {
+static Secondaries createIPSecondaries(const IPSecondariesConfig& config, bool provision, Aktualizr& aktualizr) {
   Secondaries result;
-  SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
+  if (provision) {
+    SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
 
-  for (auto& ip_sec_cfg : config.secondaries_cfg) {
-    auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
-    if (secondary) {
-      result.push_back(secondary);
-    } else {
-      sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+    for (auto& ip_sec_cfg : config.secondaries_cfg) {
+      auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
+      if (secondary) {
+        result.push_back(secondary);
+      } else {
+        sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+      }
+    }
+
+    sec_waiter.wait();
+  } else {
+    auto secondaries_info = aktualizr.GetRegisteredSecondaries();
```

Comment:
Can the secondaries that have been added to the aktualizr instance be 'not registered', if not then perhaps renaming of GetRegisteredSecondaries() to just GetSecondaries() would make sense. Or, as an option, `storage_` can be passed to createIPSecondaries instead of `aktualizr` so we can call `storage_->loadSecondaryInfo(&info);` here and replace dependency on Aktualizr with dependency on Storage which looks better since it reduces cross-dependencies, IMHO.

Reply:
`GetSecondaries()` would probably work.

My first version was also passing the storage around, the problem I had is that call originates from `src/aktualizr_primary/main.cc` directly where the underlying storage object is not accessible.

Passing `std::vector<SecondaryInfo>` might be fine or maybe too specific? Also, because of the previous point, the aktualizr method would still be needed in the current state.